File: library_zap/src/main/java/com/dolin/zap/impl/Record2MMap.java
package com.dolin.zap.impl;

import com.dolin.zap.internal.IRecord;
import com.dolin.zap.util.LogFileUtils;

import java.io.File;

/**
 * @author #Suyghur.
 * Created on 2021/05/19
 */
public class Record2MMap implements IRecord {

    private long ptr = 0L;
    private String logPath = "";
    private String logDate = "";
    private String logFilePath = "";
    private int limitSize = 0;
    private int num = 1;

    static {
        System.loadLibrary("dolin-zap");
    }

    public Record2MMap(String bufferPath, String logPath, String logDate, int capacity, int limitSize, boolean compress) {
        this.num = LogFileUtils.getLogFileNumByDate(logPath, logDate);
        this.logPath = logPath;
        this.logDate = logDate;
        this.limitSize = limitSize;
        if (num > 1) {
            this.logFilePath = logPath + File.separator + logDate + "-p" + num + ".zap";
        } else {
            this.logFilePath = logPath + File.separator + logDate + ".zap";
        }
        try {
            this.ptr = initNative(bufferPath, logFilePath, capacity, limitSize, compress);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void write(String msg) {
        if (ptr != 0L) {
            try {
                writeNative(ptr, msg);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void asyncFlush() {
        if (ptr != 0L) {
            try {
                //自动扩容
                if (isLogFileOverSizeNative(ptr)) {
                    this.num += 1;
                    this.logFilePath = logPath + File.separator + logDate + "-p" + num + ".zap";
                    expLogFileNative(ptr, logFilePath, limitSize);
                }
                asyncFlushNative(ptr);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void expLogFile(String path, int limitSize) {
        if (ptr != 0L) {
            try {
                expLogFileNative(ptr, path, limitSize);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void release() {
        if (ptr != 0L) {
            try {
                releaseNative(ptr);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private static native long initNative(String bufferPath, String logFilePath, int capacity, int limitSize, boolean compress);

    private native void writeNative(long ptr, String msg);

    private native void asyncFlushNative(long ptr);

    private native void expLogFileNative(long ptr, String path, int limitSize);

    private native void releaseNative(long ptr);

    private native boolean isLogFileOverSizeNative(long ptr);
}

File: library_zap/src/main/java/com/dolin/zap/impl/Record2MMap.kt
package com.dolin.zap.impl

import com.dolin.zap.internal.IRecord
import com.dolin.zap.util.LogFileUtils

/**
 * @author #Suyghur.
 * Created on 4/7/21
 */
class Record2MMap(private val folderPath: String, private val logDate: String,
                  capacity: Int, private val limitSize: Int, compress: Boolean) : IRecord {

    //buffer指针
    private var ptr = 0L
    private var bufferPath: String
    private var logPath: String
    private var num = 1

    init {
        System.loadLibrary("dolin-zap")
        num = LogFileUtils.getLogFileNumByDate(folderPath, logDate)
        bufferPath = "$folderPath/zap.cache"
        logPath = if (num > 1) {
            "$folderPath/$logDate/$logDate-p$num.zap"
        } else {
            "$folderPath/$logDate/$logDate.zap"
        }
        try {
            ptr = initNative(bufferPath, logPath, capacity, limitSize, compress)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun write(msg: String) {
        if (ptr != 0L) {
            try {
                writeNative(ptr, msg)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun asyncFlush() {
        if (ptr != 0L) {
            try {
                //自动扩容
                if (isLogFileOverSizeNative(ptr)) {
                    logPath = "$folderPath/$logDate/$logDate-p${num + 1}.zap"
                    expLogFileNative(ptr, logPath, limitSize)
                }
                asyncFlushNative(ptr)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun expLogFile(path: String, limitSize: Int) {
        if (ptr != 0L) {
            try {
                expLogFileNative(ptr, path, limitSize)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun release() {
        if (ptr != 0L) {
            try {
                releaseNative(ptr)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private external fun writeNative(ptr: Long, msg: String)

    private external fun asyncFlushNative(ptr: Long)

    private external fun expLogFileNative(ptr: Long, path: String, limitSize: Int)

    private external fun releaseNative(ptr: Long)

    private external fun isLogFileOverSizeNative(ptr: Long): Boolean

    companion object {
        @JvmStatic
        private external fun initNative(bufferPath: String, logPath: String, capacity: Int, limitSize: Int, compress: Boolean): Long
    }
}